File: modules/common/src/main/java/com/hig/exceptions/BaseException.java
package com.hig.exceptions;

import com.hig.types.CommonResult;
import com.hig.types.Result;
import com.hig.utils.MessageUtil;
import lombok.Getter;

/**
 * Root Exception
 */
public class BaseException extends RuntimeException {

	private static final long serialVersionUID = 3857013638852527498L;
	
	@Getter
	private final Result result;
	
	@Getter
	private final String message;
	
	public BaseException() {
		super(CommonResult.ERROR.getMessage(MessageUtil::getMessage, CommonResult.ERROR.getMessageId()));
		this.result = CommonResult.ERROR;
		this.message = CommonResult.ERROR.getMessage();
	}
	
	public BaseException(String message) {
		super(message);
		this.result = CommonResult.ERROR;
		this.message = message;
	}
	
	public BaseException(String message, Throwable cause) {
		super(message, cause);
		this.result = CommonResult.ERROR;
		this.message = message;
	}
	
	public BaseException(Throwable cause) {
		super(cause);
		this.result = CommonResult.ERROR;
		this.message = CommonResult.ERROR.getMessage();
	}

	public BaseException(Result result) {
		super(result.getMessage(MessageUtil::getMessage, result.getMessageId()));
		this.result = result;
		this.message = result.getMessage(MessageUtil::getMessage, result.getMessageId());
	}
	
	public BaseException(Result result, String message) {
		super(message);
		this.result = result;
		this.message = message;
	}

	public BaseException(Result result, Throwable cause) {
		super(result.getMessage(MessageUtil::getMessage, result.getMessageId()), cause);
		this.result = result;
		this.message = result.getMessage(MessageUtil::getMessage, result.getMessageId());
	}
	
    protected BaseException(Result result, Throwable cause, boolean enableSuppression, boolean writableStackTrace) {
    	super(result.getMessage(MessageUtil::getMessage, result.getMessageId()), cause, enableSuppression, writableStackTrace);
		this.result = result;
		this.message = result.getMessage(MessageUtil::getMessage, result.getMessageId());
    }
}

File: modules/common/src/main/kotlin/com/media/bus/common/exceptions/BaseException.kt
package com.media.bus.common.exceptions

import com.media.bus.common.result.Result
import com.media.bus.common.result.type.CommonResult

/**
 * ## Root Exception
 *
 * Kotlin 기본 인수를 활용하여 단일 주생성자로 모든 조합을 지원한다.
 * result와 message는 이 클래스에서만 선언하며, 하위 클래스는 중복 선언하지 않는다.
 */
open class BaseException(
    val result: Result = CommonResult.ERROR,
    message: String? = null,
    cause: Throwable? = null,
) : RuntimeException(
    message ?: result.message,
    cause,
) {
    final override val message: String =
        message ?: result.message
}
